walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -1,23 +1,70 @@
+    @return a Collection that has all the possible timeRanges for te requested meeting.
+  */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    
+    long meetingDuration = request.getDuration();      // This is the duration of the request meeting
+    Collection<String> meetingAttendees = request.getAttendees();    // This is a collection of strings that have the names of all the attendees in the to be set meeting
+
+    if ( (meetingDuration <= TimeRange.WHOLE_DAY.duration()) == false)
+      return Arrays.asList();
+
+    else if(events.isEmpty()) 
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    
+    // This ArrayList contains all the unavailable times of TimeRanges at which the request meeting cannot be set. 
+    // This is calculated or set by taking each event from the collection, and checking if the event attendees are busy in the collection
+    // of meetings, and hence, can't be available for the meeting request.  
+    ArrayList<TimeRange> unavailableTimes = new ArrayList<TimeRange>();
+
+    for (Event event : events){
+      Collection<String> eventAttendees = event.getAttendees();
+      TimeRange rangeOfEvent = event.getWhen();
+     
+      for (String attendee : meetingAttendees) {    // Checking the attendees of the meeting request 
+        if (eventAttendees.contains(attendee)){
+          unavailableTimes.add(rangeOfEvent);        // Comes here only if even one of the members of the request cannot make it    
+          break;                                    // So immediately exits
+        } 
+      }
+    }
+
+    Collections.sort(unavailableTimes, TimeRange.ORDER_BY_START);
+
+    // At this time, we have a list of timeRanges in unavailableTimes where the meeting cannot be set that's is sorted by order of start time. 
+    // Now let us traverse the timeranges where the meeting can be set within the working hours. 
+
+    Collection<TimeRange> availableTimes = new ArrayList<TimeRange>();  // This will have the times that have a collection of all the available timeranges of events
+    int progress = TimeRange.START_OF_DAY;            // This variable behaves like an iterator throughout the day. So, it starts from 0, which is the START time.  
+    
+    for(TimeRange rangeOfEvent : unavailableTimes){
+      int rangeEnd = rangeOfEvent.end();
+      int rangeStart = rangeOfEvent.start();
+        if (progress + meetingDuration <= rangeStart){       // If there can be a possible meeting with the requested meeting duration and that starts from progress, then it's valid.
+          TimeRange availableRange = TimeRange.fromStartEnd(progress, rangeStart, false);
+          availableTimes.add(availableRange);
+        }
+      
+      progress = Math.max(rangeEnd,progress);       // Set progress to be after the end of either events
```

Comment:
By doing this you skipped some of the possible meeting times.

If progress is 1 pm, meeting duration is 30 minutes, rangeStart is 3 pm and rangeEnd is 4 pm. You will add 1 - 1:30 and then set the progress directly to 4, so you are missing 1:30 - 2, 2 - 2:30, and 2:30 - 3.

Reply:
I don't think this is ture. TimeRange.fromStartEnd(progress, rangeStart, false) should add the time slot from 1-3pm in this case.

Reply:
Yeah you are right, I misread that line.